README.md
# Yarr!

[![Circle CI](https://img.shields.io/circleci/project/bluecap-se/yarr.svg?style=flat-square)](https://circleci.com/gh/bluecap-se/yarr)
[![Dependency Status](https://img.shields.io/gemnasium/bluecap-se/yarr.svg?style=flat-square)](https://gemnasium.com/bluecap-se/yarr)
[![NPM version](https://img.shields.io/npm/v/yarr.svg?style=flat-square)](https://www.npmjs.com/package/yarr)
[![Docker pulls](https://img.shields.io/docker/pulls/bluecap/yarr.svg?style=flat-square)](https://registry.hub.docker.com/u/bluecap/yarr/)


*I feel lucky*, for [The Pirate Bay](https://thepiratebay.se/). A micro HTTP service,
for picking the best torrent available for download, by searching TPB.

## Install

Yarr! can be installed as a package either via NPM or Docker.

#### Using Docker

```console
$ docker run -d -p 80:8080 bluecap/yarr:latest
```

#### Using a package manager

```console
$ npm install yarr
```

## Usage

```console
$ yarr
Server started on localhost:8080
```

This will start the HTTP server, with [default config](https://github.com/bluecap-se/yarr/blob/master/lib/defaults.json).
Then navigate to [localhost:8080](http://localhost:8080)

### Full options

```console
$ yarr -h

I feel lucky, for The Pirate Bay

Usage:
    yarr [<config>]
    yarr [options]

Options:
    --host=HOST       Set server host
    --port=PORT       Set server port
    --source=URL      Set The Pirate Bay URL
    --api-key=TOKEN   Set API-key
    -h --help         Show this screen
    -v --version      Show version

Examples:
       yarr (run the server with default config)
       yarr ~/config.json
       yarr --host localhost --port 8080
```

#### Using Docker

Options can be set as environment variables. Only `source` and `api-key` can be set, the rest
can be configured by using the [port option](https://docs.docker.com/reference/commandline/run/).

```console
$ docker run -d -p 80:8080 -e API_KEY='secret' -e SOURCE='http://thepiratebay.se' bluecap/yarr:latest
```

## Use case

#### Getting the best torrent

Here, [HTTPie](https://github.com/jakubroztocil/httpie) is used when calling the API over HTTP. Using Curl is another option.

```console
$ http localhost:8080/search q==Game s==1 e==5 hd==720p
{
    "name": "...",
    "added": "05-16 2011",
    "magnet": "magnet:?xt=urn:...",
    "size": "1.46 GiB",
    "seeders": "69",
    "leachers": "10"
}
```

#### Getting the magnet

Here, [jq](http://stedolan.github.io/jq/) is used to parse the result from Yarr!. The result can then be piped to a torrent application of your choice.

```console
$ http localhost:8080/search q==Game s==1 e==5 hd==720p | jq .magnet
magnet:?xt=urn:
```

## API

All response is JSON formatted.

#### Perform a search
Returned is the best matching torrent.

```console
$ http localhost:8080/search q==value [param==value]
```

<table>
    <tr>
        <th>Param</th>
        <th>Required</th>
        <th>Description</th>
        <th>Input type</th>
    </tr>
    <tr>
        <td>q</td>
        <td>Yes</td>
        <td>Search query.</td>
        <td>Freetext string</td>
    </tr>
    <tr>
        <td>s</td>
        <td>No</td>
        <td>Season, for shows.</td>
        <td>Number</td>
    </tr>
    <tr>
        <td>e</td>
        <td>No</td>
        <td>Episode, for shows. Requires season.</td>
        <td>Number</td>
    </tr>
    <tr>
        <td>hd</td>
        <td>No</td>
        <td>Flag for HD.</td>
        <td>720p, 1080p</td>
    </tr>
</table>

#### Check version
Returned is the running version of Yarr!

```console
$ http localhost:8080/version
{
    "version": "1.1.0"
}
```

## Run tests

```console
$ make test
```

## License

Published under [MIT License](https://github.com/bluecap-se/yarr/blob/master/LICENSE).


HISTORY.md
# Changelog


## 1.1.0

- Added docker
- Changed from Travis CI to Circle CI

## Minor: 1.0.2

- NPM publishing tweaks

## Minor: 1.0.1

- Repo renaming

## 1.0.0

- First release. Minimal viable product
- Server and API
- Fetching and parsing from TPB
- Config parser
- Test suite


test/parser.js
var expect = require('chai').expect,
    parser = require('../lib/parser'),
    readFile = require('./support/readfile');


describe('Parser', function() {
    var file = __dirname + '/fixtures/thepiratebay.html',
        HTML = readFile(file),
        p = new parser(HTML);

    describe('list', function() {
        it('should have 14 rows', function() {
            var a = p.parse();
            expect(p.data.length).to.equal(14);
        });
    });
});


test/server.js
var request = require('supertest'),
    expect = require('chai').expect,
    config = require('../lib/config'),
    app = require('../lib/server').app,
    pkg = require('../package.json');


describe('Server', function() {

    describe('responds with JSON', function() {
        it('on /', function(done) {
            request(app)
                .get('/')
                .expect('Content-Type', /json/)
                .expect(302, done);
        });

        it('on /search', function(done) {
            request(app)
                .get('/search')
                .expect('Content-Type', /json/)
                .expect(400, done);
        });

        it('on /version', function(done) {
            request(app)
                .get('/version')
                .expect('Content-Type', /json/)
                .expect(200, done);
        });
    });

    describe('runs with API-key', function() {
        before(function() {
            config.apiKey = 'testing';
        });

        it('with invalid token', function(done) {
            request(app)
                .get('/')
                .expect(401, done);
        });

        it('from querystring', function(done) {
            request(app)
                .get('/?api-key=testing')
                .expect(302, done);
        });

        it('from header', function(done) {
            request(app)
                .get('/')
                .set('X-API-KEY', 'testing')
                .expect(302, done);
        });

        after(function() {
            config.apiKey = "";
        })
    });

    describe('GET /', function() {
        it('contains a redirection', function(done) {
            request(app)
                .get('/')
                .expect('Location', /search/)
                .expect(302, done);
        });
    });

    describe('GET /search', function() {
        it('requires a query param', function(done) {
            request(app)
                .get('/search')
                .expect(400, function(err, res) {
                    if (err) {
                        return done(err);
                    }

                    expect(res.query).to.not.exist;
                    expect(res.body.error).to.not.be.empty;
                    done();
                });

            request(app)
                .get('/search?q=test')
                .end(function(err, res) {
                    if (err) {
                        return done(err);
                    }

                    expect(res.query.q).to.exist;
                    done();
                });
        });
    });

    describe('GET /version', function() {
        it('shows correct version', function(done) {
            request(app)
                .get('/version')
                .end(function(err, res) {
                    if (err) {
                        return done(err);
                    }

                    expect(res.body.version).to.equal(pkg.version);
                    done();
                });
        });
    });

});


lib/config.js
var config = defaults = require('./defaults.json');

module.exports = new Config();


function Config() {
    this.host = config.server.host;
    this.port = config.server.port;
    this.apiKey = config.server.apiKey;
    this.source = config.tpb.url;
    this.categories = config.tpb.categories;
}

/*
    Load configuration file
*/
Config.prototype.load = function(path) {
    try {
        config = require(path);
        this.setByEval('host', 'config.server.host');
        this.setByEval('port', 'config.server.port');
        this.setByEval('apiKey', 'config.server.apiKey');
        this.setByEval('source', 'config.tpb.url');
    }
    catch(e) {
        console.error('Error: Failed to load config file, ' +
                      'defaulting to standard config...');
    }
};

/*
    Setting a config value,
    by evaluating and catching key errors
*/
Config.prototype.setByEval = function(name, value) {
    try {
        this.set(name, eval(value));
    }
    catch(e) {}
};

/*
    Setting a config value,
    default back to original
*/
Config.prototype.set = function(name, value) {
    try {
        if (!(name in this)) {
            throw 'KeyError';
        }
        this[name] = value || this[name];
    }
    catch(e) {
        console.error('Error: Config key "%s" doesn\'t exist.', name);
    }
}


lib/parser.js
var cheerio = require('cheerio');


module.exports = HTMLParser;


function HTMLParser(html) {
    if (html) {
        this.load(html);
    }

    this.data = [];
    this.types = {
        1: 'name',
        2: 'added',
        3: 'magnet',
        4: 'size',
        5: 'seeders',
        6: 'leachers'
    };
}

HTMLParser.prototype.load = function(html) {
    this.$ = cheerio.load(html);
};

HTMLParser.prototype.extract = function(index, $el) {
    switch(index) {
        case 1:
            return $el.find('a').text() || '';
            break;
        case 2:
        case 4:
        case 5:
        case 6:
            return $el.text();
            break;
        case 3:
            return $el.find('a[href^=magnet]').attr('href') || '';
    }
};

HTMLParser.prototype.cols = function($row) {
    var self = this,
        obj = {},
        key, val;

    $row.find('td').each(function(i) {
        if (i === 0) {
            return;
        }

        val = self.extract(i, self.$(this));
        key = self.types[i];
        obj[key] = val;
    });

    return obj;
};

HTMLParser.prototype.parse = function() {
    var self = this,
        obj;

    this.$('#searchResult tr:not(.header)').each(function(i) {
        obj = self.cols(self.$(this));
        self.data.push(obj);
    });
};

HTMLParser.prototype.results = function(filter) {
    return this.data;
};


lib/fetcher.js
var request = require('request'),
    cheerio = require('cheerio'),
    config = require('./config');


module.exports = Fetcher;


function Fetcher(query, hd, season, episode) {
    this.options = {
        url: config.source,
        headers: {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.63 Safari/537.36',
            'Cookie': 'lw=s'
        }
    };

    this.urlBuilder(query, hd, season, episode);
}

Fetcher.prototype.urlBuilder = function(query, hd, season, episode) {
    var url = this.options.url + '/search/%q/0/7/%cat',
        q = [query],
        hdChoices = ['720p', '1080p'],
        prependZero = function(num) {
            return (num.length === 1) ? '0' + num : num;
        };

    if (season) {
        var s = ['S', prependZero(season)];

        if (episode) {
            s = s.concat(['E', prependZero(episode)]);
        }

        q.push(s.join(''));
    }

    if (hd && hdChoices.indexOf(hd) > -1) {
        q.push(hd);
    }

    q = encodeURIComponent(q.join(' '));
    this.options.url = url.replace('%cat', config.categories.video).replace('%q', q);
};

Fetcher.prototype.load = function(callback) {
    request(this.options, callback);
};


lib/server.js
var express = require('express'),
    fetcher = require('./fetcher'),
    parser = require('./parser'),
    config = require('./config'),
    pkg = require('../package.json'),
    app = express();


var start = function() {
    var port = config.port,
        host = config.host;

    app.listen(port, host, function() {
        console.log('Server started on %s:%s', host, port);
    });
};


var sendError = function(res, status, msg) {
    var body = { 'error': msg };

    res.status(status).send(body);
    return;
};


module.exports = {
    'start': start,
    'app': app
};

/*
 * Middleware
 */

app.use(function(req, res, next) {

    // All requests are in JSON format
    res.setHeader('Content-Type', 'application/json');

    // Check for API key
    if (config.apiKey) {
        var recieved = req.query['api-key'] || req.get('X-API-KEY') || false;
        if (config.apiKey != recieved || !recieved) {
            return sendError(res, 401, 'API-key reguired, bad token supplied');
        }
    }

    next();
});


/*
 * Routes
 */

app.get('/', function(req, res) {
    res.setHeader('Location', '/search');
    sendError(res, 302, 'Welcome to Yarr! Take a look at /search');
});


app.get('/search', function(req, res) {

    // Check for q in params
    if (!req.query.q) {
        return sendError(res, 400, 'Missing required \'q\' parameter.');
    }

    var q = req.query.q,
        hd = req.query.hd || false,
        season = req.query.s || false,
        episode = req.query.e || false,
        body;

    var parseTester = function(err, _res, body) {
        if (err || _res.statusCode !== 200) {
            return sendError(res, 500, 'Couldn\'t fetch from URL.');
        }

        var a = new parser(body);
        a.parse();
        body = a.data[0] || { 'error': 'No results found.'};
        res.send(body);
    };

    var f = new fetcher(q, hd, season, episode);
    f.load(parseTester);
});


app.get('/version', function(req, res) {
    var body = { 'version': pkg.version };
    res.send(body);
});
